### PR TITLE
Fixed #54: added search result processing to add facet aggregates

### LINKED ISSUE #54: Getting aggregates for facets
Having seen showMoreFacets, we should pull a similar trick to get aggregates like min/max for facets. Would be super-useful for range sliders and date pickers for non-string facets..

### PR DESCRIPTION
**Pull request by grtjn**
Fixes #54

### CODE REVIEW

File: src/ml-search.service.js [hunk 1]
```diff
@@ -1225,6 +1233,77 @@
...
+          var searchOptions = self.getQuery();
+          searchOptions.options = {};
+          searchOptions.options.constraint = [_.cloneDeep(myOption)];
+          if (myOption.range && myOption.range['facet-option']) {
+            myOption['values-option'] = myOption.range['facet-option'];
+            
+            // these work for all index types
+            myOption.aggregate = [
+              { apply: 'count' },
+              { apply: 'min' },
```
Comment (joemfb) [inline src/ml-search.service.js]: I'm pretty sure `min`/`max` don't work for all index types. I tested pretty extensively when I was developing https://github.com/joemfb/analytics-query-browser. Here's the module I used to create the config for that project: https://gist.github.com/joemfb/b682504c7c19cd6fae11

**Comment by joemfb** [comment]
This makes sense. Let me review my values branch, check the tests, etc. Than I'll do a PR. Once that's merged, you can rebase on top of this. Should let you simplify parts of it.

File: src/ml-search.service.js (hunk 1 above)
Comment (grtjn) [inline src/ml-search.service.js]: I admit I didn't test *all* types, but I did test xs:string, xs:dateTime, and geo-elem-pair. I think that covers the most difficult ones..

Comment (grtjn) [inline src/ml-search.service.js]: Looks like you are only missing string in your list for min/max. And I'd expect anyUri to work as well..

**Comment by grtjn** [comment]
Looking forward to your PR.. :)

**Comment by grtjn** [comment]
I rebased, synched my getAggregates code with your new showMoreFacets. I ran tests against a project I was using for the new functionality, and found a few issues in values and showMoreFacets. The options need to be put in combined.search.options, and you need to cloneDeep constraints from storedOptions. Otherwise you end up messing them up..

Looks good like this? gulp tests run clean. I can squash the commits if you like..

File: src/ml-search.service.js [hunk 2]
```diff
@@ -767,13 +769,15 @@
           return $q.reject(new Error('No constraint exists matching ' + facetName));
         }
 
-        var newOptions = { constraint: constraint, values: constraint };
-        newOptions.values['values-option'] = newOptions.values.range['facet-option'];
+        constraint = _.cloneDeep(constraint);
+
+        var newOptions = { constraint: _.cloneDeep(constraint), values: constraint };
```
Comment (joemfb) [inline src/ml-search.service.js]: Are both `_.cloneDeep()` invocations necessary? The original constraint object isn't used again, so I would think one would be enough ...

**Comment by joemfb** [comment]
My only question is the `cloneDeep` calls. Otherwise, I'd say squash the commits and we can merge (I don't think they're sufficiently differentiated to be valuable separately).

Looks good!

**Comment by grtjn** [comment]
two clones was indeed overkill. I fixed that, and squashed the commits!

**Comment by joemfb** [comment]
Actually, I was wrong; there's one more thing...

Could you write some tests for the `getAggregates` method? There's several examples of tests for similar methods. You can mock the data, and then use the code coverage report to figure out where the gabs are, and what additional tests you need.

**Comment by grtjn** [comment]
I wish it was that easy as copy paste one of the other tests, but the code loops over all facets, and makes all kinds of backend calls. I'm not very good at writing tests for that. Can you give a few more hints how to best approach this?

**Comment by joemfb** [comment]
I've tried to make the test harness straightforward, but some of this code is just really hard to test. You can mock the dependencies directly, and just make sure that they're invoked with the correct parameters, or you can use `$httpBackend` to mock the underlying HTTP requests.

`showMoreFacets` is one of the more complicated sets of tests, and probably a useful example here: https://github.com/joemfb/ml-search-ng/blob/master/test/spec/ml-search.service.js#L580

Also, I mocked MLRest directly here: https://github.com/joemfb/ml-search-ng/blob/master/test/spec/ml-search.service.js#L1266. That might be a simpler approach in this case.

**Comment by grtjn** [comment]
Thanks for the hints, I added two test cases for getAggregates. I also rebased, and fixed the test cases for showMoreFacets (issue #45). Additionally, jshint was giving some suggestions at my end, to good use I think..

I'll squash commits if you like the changes..

**Comment by joemfb** [comment]
This looks good, thanks for adding the test cases!

Please do rebase down to one commit, and, when you do, can you remove `dist/*`? I want those to change only with tagged release commits.

**Comment by grtjn** [comment]
Whoops, didn't mean to commit dist. I have squashed the commits..

File: src/ml-search.service.js [hunk 3]
```diff
@@ -1137,16 +1141,16 @@
       }
 
       params = params || {};
-      params.start = params.start || 1;
-      params.limit = params.limit || 20;
+      params.start = params.start !== undefined ? params.start : 1;
+      params.limit = params.limit !== undefined ? params.limit : 20;
```
Comment (joemfb) [inline src/ml-search.service.js]: This more explicit check is to allow falsy values as parameters (such as `limit=0`)?

Comment (grtjn) [inline src/ml-search.service.js]: Yes exactly. The getAggregates passes through limit=0 to only get aggregates, and not bother about values. My test case was complaining about it.. ;-)

**Pull request merged**